### DISCUSSION TITLE
Upcoming Changes

### DISCUSSION THREAD

**Discussion by MGatner** [tattersoftware/codeigniter4-settings#31] [Announcements]
Since this library's last major version the CodeIgniter team has published its own
[Settings Library](https://github.com/codeigniter4/settings). That library and this offer
some non-overlapping features but they do represent a conflict in name and implementation.
The future of this library is not yet determined, but some options are:
* Deprecate this library and move any missing features to the CodeIgniter version
* Rework this library as an extension of the official package
* Maintain the current functionality with a name change to avoid conflict

**Comment by sfadschm** [comment]
That's an interesting development.
While I don't yet overlook the feature differences between the implementations, I think merging additional features of the `tatter` version into the `ci4` version sounds most reasonably. Of course only if they are desired there.

**Reply by MGatner** [reply]
I think I have settled on my solution. This library will cease to exist. The official `Settings` library will be expanded to include "contextual settings", and I will release the missing features from this library as a new package (https://github.com/tattersoftware/codeigniter4-preferences).

**Reply by sfadschm** [reply]
Alright, curious how this will turn out 😃

**Reply by MGatner** [reply]
I'm glad you found the main PR - for anyone else who comes along: https://github.com/codeigniter4/settings/pull/16

I will post an Upgrade Guide to this repo once it is deprecated and the new packages are ready, but it should be fairly easy. The biggest conceptual change is that everything is Config based now, so while you can technically create a Setting without a corresponding Config file it is a good idea to add one (I usually name mine "Project"). After that global settings will be accessed via the Setting helper (e.g. `setting('Project.appName')` and user-specific settings will use the Preference helper (e.g. `preference('Project.theme')`).

**Comment by MGatner** [comment]
`Tatter\Preferences` is now published in its first release candidate form:
* https://packagist.org/packages/tatter/preferences
* https://github.com/tattersoftware/codeigniter4-preferences
* https://forum.codeigniter.com/thread-80597.html

I will be provided an upgrade guide in this repo and deprecate that classes so folks are aware, but for anyone who has made their way to this thread I recommend checking out the new material ASAP.

**Comment by sfadschm** [comment]
Preferences looks like a nice addition to the new Settings.
Thanks for all the work!
Although I do not need the Preferences extension right now, it looks nicely focussed and lightweight 😄

**Reply by MGatner** [reply]
Thanks @sfadschm! Curiously I'm finding as I rewrite some of these older libraries to find they are getting much smaller, a trend I like. 

I have one more change for this library for (hopefully) a last RC then we will take it to production.
